client/src/main/java/org/eclipse/hono/client/impl/TenantClientImpl.java

```diff
@@ -0,0 +1,155 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.client.impl;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.proton.ProtonReceiver;
+import io.vertx.proton.ProtonSender;
+import org.eclipse.hono.client.StatusCodeMapper;
+import org.eclipse.hono.client.TenantClient;
+import org.eclipse.hono.config.ClientConfigProperties;
+
+import org.eclipse.hono.util.TenantConstants;
+import org.eclipse.hono.util.TenantObject;
+import org.eclipse.hono.util.TenantResult;
+import org.eclipse.hono.util.MessageHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonConnection;
+
+/**
+ * A Vertx-Proton based client for Hono's Tenant API.
+ *
+ */
+public final class TenantClientImpl extends AbstractRequestResponseClient<TenantResult<TenantObject>>
+        implements TenantClient {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TenantClientImpl.class);
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private TenantClientImpl(final Context context, final ClientConfigProperties config, final String tenantId) {
+        super(context, config, tenantId);
+    }
+
+    TenantClientImpl(final Context context, final ClientConfigProperties config, final String tenantId,
+                           final ProtonSender sender, final ProtonReceiver receiver) {
+
+        super(context, config, tenantId, sender, receiver);
+    }
+
+    @Override
+    protected String getName() {
+
+        return TenantConstants.TENANT_ENDPOINT;
+    }
+
+    @Override
+    protected String createMessageId() {
+
+        return String.format("%s-%s", TenantConstants.MESSAGE_ID_PREFIX, UUID.randomUUID());
+    }
+
+    @Override
+    protected TenantResult<TenantObject> getResult(final int status, final String payload) {
```

Comment:
Can you make all the existing methods to be final, so that any extensions of the `TenantClientImpl` doesn't modify them.

Reply:
Of course, when I removed the final from the class declaration, I introduced final for the methods instead. Did not push it yet ;-)